### PR TITLE
Adding API's to enable lazy tensor initialization

### PR DESCRIPTION
**Pull request by gautham-kollu**
Adding API's to enable lazy tensor initialization

## What does this PR do?
Enable lazy initialization of tensors.  This is the first commit to allow lazy initialization of larger models that would creation of models that don't even fit into one GPU.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by anj-s** [comment]
Can you add a unit test for this PR?

File: fairscale/utils/meta.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
+####################################################################
+# BELOW: TAKEN FROM Pytorch-lightning. https://github.com/PyTorchLightning/pytorch-lightning/pull/9920#
+# Pruned part of the code dependent on Pytorch-lightning
+# TODO: Removed once merged and released on Pytorch-lightning side #
+####################################################################
+
+# Copyright The PyTorch Lightning team.
```
Comment (anj-s) [inline fairscale/utils/meta.py:7]: We can use the FairScale copyright as long as we give credit to the PTL folks.

Comment (anj-s) [reply fairscale/utils/meta.py:7]: I also remember there being some bug fixes so might be best if we take the latest code as opposed to something from the PR.

**Comment by cbalioglu** [comment]
As I mentioned offline on Friday, I strongly encourage you to start using the implementation in torchdistx since this version copied from PyTorch Lightning has many shortcomings.

**Comment by anj-s** [comment]
> As I mentioned offline on Friday, I strongly encourage you to start using the implementation in torchdistx since this version copied from PyTorch Lightning has many shortcomings.

Is the version from torchdistx available as PyPI package that we can install?

**Pull request closed without merging**